concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,1481 @@
+    /// - `children[i]` is a subtree containing strictly smaller keys.
+    /// - `children[i + 1]` is a subtree containing strictly larger keys.
+    children: Vec<NodeId>,
+}
+
+/// Wrapper implement the exact same deserial as K, but wraps it in an
+/// option in memory. This is used, to allow taking a key from a mutable
+/// reference to a node, without cloning the key, during iteration of the
+/// set.
+#[repr(transparent)]
+struct KeyWrapper<K> {
+    key: Option<K>,
+}
+
+impl<K: Deserial> Deserial for KeyWrapper<K> {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        let key = K::deserial(source)?;
+        Ok(Self {
+            key: Some(key),
+        })
+    }
+}
+
+/// Byte size of the key used to store a BTree internal node in the smart
+/// contract key-value store.
+const BTREE_NODE_KEY_SIZE: usize = STATE_ITEM_PREFIX_SIZE + NodeId::SERIALIZED_BYTE_SIZE;
+
+impl<const M: usize, K> Node<M, K> {
+    /// The max length of the child list.
+    const MAXIMUM_CHILD_LEN: usize = 2 * M;
+    /// The max length of the key list.
+    const MAXIMUM_KEY_LEN: usize = Self::MAXIMUM_CHILD_LEN - 1;
+    /// The min length of the child list, when the node is not a leaf node.
+    const MINIMUM_CHILD_LEN: usize = M;
+    /// The min length of the key list, except when the node is root.
+    const MINIMUM_KEY_LEN: usize = Self::MINIMUM_CHILD_LEN - 1;
+
+    /// The number of keys stored in this node.
+    fn len(&self) -> usize { self.keys.len() }
+
+    /// Check if the node holds the maximum number of keys.
+    fn is_full(&self) -> bool { self.len() == Self::MAXIMUM_KEY_LEN }
+
+    /// Check if the node is representing a leaf in the tree.
+    fn is_leaf(&self) -> bool { self.children.is_empty() }
+
+    /// Check if the node holds the minimum number of keys.
+    fn is_at_min(&self) -> bool { self.len() == Self::MINIMUM_KEY_LEN }
+}
+
+impl NodeId {
+    /// Byte size of `NodeId` when serialized.
+    const SERIALIZED_BYTE_SIZE: usize = 4;
+
+    /// Return a copy of the NodeId, then increments itself.
+    fn copy_then_increment(&mut self) -> Self {
+        let current = *self;
+        self.id += 1;
+        current
+    }
+
+    /// Construct the key for the node in the key-value store from the node ID.
+    fn as_key(&self, prefix: &StateItemPrefix) -> [u8; BTREE_NODE_KEY_SIZE] {
+        // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
+        // safe because the type we are claiming to have initialized here is a
+        // bunch of `MaybeUninit`s, which do not require initialization.
+        let mut prefixed: [mem::MaybeUninit<u8>; BTREE_NODE_KEY_SIZE] =
+            unsafe { mem::MaybeUninit::uninit().assume_init() };
+        for i in 0..STATE_ITEM_PREFIX_SIZE {
+            prefixed[i].write(prefix[i]);
+        }
+        let id_bytes = self.id.to_le_bytes();
+        for i in 0..id_bytes.len() {
+            prefixed[STATE_ITEM_PREFIX_SIZE + i].write(id_bytes[i]);
+        }
```

Comment:
```suggestion
        for (place, value) in prefixed[STATE_ITEM_PREFIX_SIZE..].iter_mut().zip(id_bytes) {
            place.write(value);
        }
```

Reply:
I'm also a bit confused as to why we do this here. Is it worse to allocate a [0u8; _] and just copy-from-slice?

Reply:
Not sure that I follow.
Do you mean avoid using `MaybeUninit` completely?

Reply:
Yes.

Reply:
I'm wondering if there will actually be a difference.

Reply:
A comparison with and without should settle that. I'll note this down when I get to benchmarking again

Reply:
I compared these and it does improve it. Very little, but still an improvement